File: ios-native/NeobankNative/Sources/Features/Cards/CardsController.swift
import Foundation
import Observation

@MainActor
@Observable
final class CardsController {
    enum LoadState {
        case loading
        case loaded([BankCard])
        case failed(APIError)
    }

    private(set) var state: LoadState = .loading

    private let repository: CardRepository

    init(repository: CardRepository) {
        self.repository = repository
    }

    func load() async {
        state = .loading
        do {
            state = .loaded(try await repository.list())
        } catch let error as APIError {
            state = .failed(error)
        } catch {
            state = .failed(.decoding())
        }
    }

    func issueCard(cardholderName: String, dailyLimit: String?, onlineOnly: Bool?) async throws {
        _ = try await repository.issue(cardholderName: cardholderName, dailyLimit: dailyLimit, onlineOnly: onlineOnly)
        await load()
    }

    func toggleFreeze(_ card: BankCard) async throws {
        let updated = card.isFrozen ? try await repository.unfreeze(card.id) : try await repository.freeze(card.id)
        guard case .loaded(var cards) = state, let index = cards.firstIndex(where: { $0.id == updated.id }) else { return }
        cards[index] = updated
        state = .loaded(cards)
    }
}

#if DEBUG
extension CardsController {
    static var preview: CardsController {
        CardsController(repository: CardRepository(client: APIClient(baseURL: AppConfig.apiBaseURL)))
    }
}
#endif

File: android-native/app/src/main/kotlin/dev/horobets/neobanknative/features/cards/CardsController.kt
package dev.horobets.neobanknative.features.cards

import androidx.compose.runtime.staticCompositionLocalOf
import dev.horobets.neobanknative.core.networking.ApiError
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

sealed interface CardsLoadState {
    data object Loading : CardsLoadState
    data class Loaded(val cards: List<BankCard>) : CardsLoadState
    data class Failed(val error: ApiError) : CardsLoadState
}

class CardsController(private val repository: CardRepository) {
    private val _state = MutableStateFlow<CardsLoadState>(CardsLoadState.Loading)
    val state: StateFlow<CardsLoadState> = _state.asStateFlow()

    suspend fun load() {
        _state.value = CardsLoadState.Loading
        _state.value = try {
            CardsLoadState.Loaded(repository.list())
        } catch (e: ApiError) {
            CardsLoadState.Failed(e)
        } catch (e: Exception) {
            CardsLoadState.Failed(ApiError.decoding())
        }
    }

    suspend fun issueCard(cardholderName: String, dailyLimit: String?, onlineOnly: Boolean?) {
        repository.issue(cardholderName, dailyLimit, onlineOnly)
        load()
    }

    suspend fun toggleFreeze(card: BankCard) {
        val updated = if (card.isFrozen) repository.unfreeze(card.id) else repository.freeze(card.id)
        val loaded = _state.value as? CardsLoadState.Loaded ?: return
        _state.value = CardsLoadState.Loaded(loaded.cards.map { if (it.id == updated.id) updated else it })
    }
}

val LocalCardsController = staticCompositionLocalOf<CardsController> {
    error("No CardsController provided — wrap the composition in AppEnvironment's providers")
}
